elements-fun/src/transaction.rs

```diff
@@ -330,16 +547,167 @@ pub struct PegoutData<'txo> {
     pub extra_data: Vec<&'txo [u8]>,
 }
 
+// TODO: think about tagging
 /// Transaction output
-#[derive(Clone, Default, PartialEq, Eq, Debug, Hash)]
-#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize), serde(crate = "serde_crate"))]
-pub struct TxOut {
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(crate = "serde_crate")
+)]
+pub enum TxOut {
+    Null(NullTxOut),
+    Explicit(ExplicitTxOut),
+    Confidential(ConfidentialTxOut),
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(crate = "serde_crate")
+)]
+pub struct NullTxOut {
+    pub script_pubkey: Script,
+}
+
+impl NullTxOut {
+    pub fn encoded_length(&self) -> usize {
+        1 + 1
+            + 1
+            + VarInt(self.script_pubkey.len() as u64).len() as usize
+            + self.script_pubkey.len()
+    }
+}
+
+impl Encodable for NullTxOut {
+    fn consensus_encode<S: io::Write>(&self, mut s: S) -> Result<usize, encode::Error> {
+        Ok(0u8.consensus_encode(&mut s)?
+            + 0u8.consensus_encode(&mut s)?
+            + 0u8.consensus_encode(&mut s)?
+            + self.script_pubkey.consensus_encode(&mut s)?)
+    }
+}
+
+impl Decodable for NullTxOut {
+    fn consensus_decode<D: io::BufRead>(mut d: D) -> Result<NullTxOut, encode::Error> {
+        let asset_tag = u8::consensus_decode(&mut d)?;
+        let value_tag = u8::consensus_decode(&mut d)?;
+        let nonce_tag = u8::consensus_decode(&mut d)?;
+
+        debug_assert_eq!(asset_tag, 0);
+        debug_assert_eq!(value_tag, 0);
+        debug_assert_eq!(nonce_tag, 0);
+
+        Ok(NullTxOut {
+            script_pubkey: Decodable::consensus_decode(&mut d)?,
+        })
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(crate = "serde_crate")
+)]
+pub struct ExplicitTxOut {
+    /// Committed asset
+    pub asset: ExplicitAsset,
+    /// Committed amount
+    pub value: ExplicitValue,
+    /// Scriptpubkey
+    pub script_pubkey: Script,
+}
+
+impl ExplicitTxOut {
+    pub fn encoded_length(&self) -> usize {
+        self.asset.encoded_length()
+            + self.value.encoded_length()
+            + 1 // nonce
+            + VarInt(self.script_pubkey.len() as u64).len() as usize
+            + self.script_pubkey.len()
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash, PartialOrd, Ord)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(crate = "serde_crate", transparent)
+)]
+pub struct ExplicitValue(pub u64);
```

Comment:
:question: Should we be using `bitcoin::Amount` under the hood? Is that a horrible idea?

Reply:
Not a horrible idea because all tokens have the same decimal points :)